### ISSUE TITLE
disallow applying same discount code to a cart more than once

### ISSUE THREAD

**Issue by bayareacoder** [reactioncommerce/api-plugin-discounts-codes#7]
It seems there is no protection against a customer applying the same discount code more than once to a cart. So he can receive the same discount multiple times... 

Solution: add a check in function `applyDiscountCodeToCart` to check if the same `discountId` is already present in `cart.billing[]`. If so, do not add it again.

**Comment by karaarda** [comment]
Adding the transaction after the order is created is not a good option. It leads to an exploit were more than one user adds the discount code to their carts exceeding the usage limits of the discount. There needs to be another check then placing the order to prevent this exploit, which leads to a more complex code base and a confusing experience for the users.

Transaction should be added when the discount is applied to the cart as follows:

```
  const transaction = { cartId, userId, appliedAt };

  // Instead of directly updating cart, we add the discount billing
  // object from the existing cart, then pass to `saveCart`
  // to re-run cart through all transforms and validations.
  const savedCart = await context.mutations.saveCart(context, cart);
  await Discounts.updateOne({
    _id: discount._id
  }, {
    $addToSet: {
      transactions: transaction
    }
  });
```

and it needs to be removed when the discount code is removed from the cart as follows:

```
  const transactionToRemove = discount.transactions.filter((transaction) => transaction.cartId === cartId);

  await Discounts.updateOne({
    _id: discountId,
    shopId
  }, {
    $pull: {
      transactions: transactionToRemove
    }
  });
```

and the `afterOrderCreate` event should be removed.

I didn't test the code yet so it might have problems. @bayareacoder let me know if there are any problems.

**Comment by bayareacoder** [comment]
_"Adding the transaction after the order is created is not a good option. It leads to an exploit were more than one user adds the discount code to their carts exceeding the usage limits of the discount."_

I'm not sure I see this as an issue. The current implementation checks, at the time of applying a discount code to a cart, that the limits are not exceeded, based on actually placed orders. It is only when the order is placed that counts as a 'usage' of the discount. Since discounts and orders are in separate plugins, the only way to communicate between both is via app events and it happens immediately after order is placed.

The issue I mentioned is that there is no check currently for the same user applying the discount more than once on the same order. He should not be able to add any given discount code more than once to a cart.

**Comment by karaarda** [comment]
Yes, I understood the issue you mentioned. However; imagine that a discount code is limited to be used only once and it is added to four carts with success. If they try to place an order they will all be able to use the discount code because there is no check for that. Also, carts have an optional expire date; hence, when a cart is abandoned it shouldn't be counted towards the limit once it expires.

**Comment by bayareacoder** [comment]
OK. In the interest of avoid too much complexity I think the current solution is good enough for typical scenarios since code is only entered at the end of the checkout process. So the time between applying code to cart and placing order is short, so the chance of many customers entering it at same time is low.
It's also not likely you would promote a code that only can be used by 1 or a few customers (would lead to frustration with many shoppers who cannot use it). More likely you'll have something for the 'first 100' customers, and then it doesn't matter much if it's used 100 or 101,102 times. 
But I agree in a "rush sale" scenario, when everyone could be ordering at the same time, it could lead to significant more usage.

**Comment by karaarda** [comment]
Yes, I was addressing to that bad UX, hence adding the usage at the time the code is used is better I believe but it is a matter of taste. However, customers does not have to place an order right after using the code, they can come back later to make the purchase, or they can continue adding new items however they please. Nevertheless; we have implemented our own discount plugin as it is not certain when they will get back to developing reaction.

**Comment by bayareacoder** [comment]
Just curious... how do you detect an abandoned shopping cart to remove the discount transaction if you record a discount transaction as soon as they apply the discount to their cart? What if they just close their browser prior to placing order?

**Comment by karaarda** [comment]
As I have mentioned there is an expiration date on the cart objects. If the cart is not used to place an order before the expiration date, then you can assume that the discount is not used.

**Comment by bayareacoder** [comment]
I get that but the cart is deleted by the server code of the storefront at expiration. Haven't looked into it but i assume it's just a server cookie that expires. 
How do you go from that to removing the discount transaction on Reaction?

**Comment by karaarda** [comment]
We didn't implement that part yet as we are not concerned with that case as much as you however you can follow several approaches. I am not sure what purpose cart expiration date serves, it might be only used for anonymous carts but you can easily tweak it to use for other purposes. As workarounds you can add an event for cart expiration, you can filter the transactions whose cart are expired and has no related order placed etc.. Let me know if you have other questions, we can discuss in private on another platform.

**Comment by bayareacoder** [comment]
Ok. So your proposal is not really fully worked out yet. No need to discuss more. Thank you for the clarification.
